keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,477 @@
+2. Check mount options of the volume mount dir. If and only if it contains `seclabel` mount option, the volume supports SELinux.
+
+### Goals
+
+Optionally (chosen by user), do not recursively relabel content of the volumes.
+
+### Non-Goals
+
+Change container runtimes / CRI.
+
+## Proposal
+
+Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+
+```go
+type SELinuxRelabelPolicy string
+
+const (
+    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // SELinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+
+    // For context:
+    // fsGroupChangePolicy defines behavior of changing ownership and permission of the volume
+    // before being exposed inside Pod. This field will only apply to
+    // volume types which support fsGroup based ownership(and permissions).
+    // It will have no effect on ephemeral volume types such as: secret, configmaps
+    // and emptydir.
+    // Valid values are "OnRootMismatch" and "Always". If not specified defaults to "Always".
+    // +optional
+    FSGroupChangePolicy *PodFSGroupChangePolicy `json:"fsGroupChangePolicy,omitempty" protobuf:"bytes,9,opt,name=fsGroupChangePolicy"`
+
+    ...
+}
+```
+
+See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-skip-permission-change.md for similar API for ownership change for fsGroup.
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (`mount` here vs. recursive `chown` in the other KEP).
+
+In order to allow `SELinuxRelabelPolicy: OnVolumeMount` for volumes provided by CSI drivers, kubelet must know if a CSI driver supports SELinux or not.
+
+```go
+// In storage.k8s.io/v1:
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // supports SELinux specifies that volumes provided by the the CSI driver support
+    // filesystem extended attributes and SELinux. Typically, volumes based on a block
+    // device support SELinux and volumes based on a shared filesystem (NFS, CIFS, CephFS...)
+    // don't.
+    //
+    // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+    // option for volumes of a pod with
+    // podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount" or the container runtime
+    // may change label of all files on the volume to match the Pod requirements.
```

Comment:
Does the scenario where PVC shared between 2 pods ( with different podSecurityContext.seLinuxRelabelPolicy) scheduled on the same node covered here?